tfhe/src/integer/u256.rs

```diff
@@ -9,6 +9,90 @@ pub const fn adc(l: u64, r: u64, c: bool) -> (u64, bool) {
+        // so, if target_endian == big it means, the full range
+        // B0..B31 is not a true little endian order,
+        // thus requiring an additional step
+
+        let inner_slice = self.0.as_mut_slice();
+        inner_slice.fill(0);
+        unsafe {
+            // SAFETY SLICES are contiguous
+            // (4 * 64) bits / 8 bits = 32
+            let inner_byte_slice =
+                std::slice::from_raw_parts_mut(inner_slice.as_mut_ptr() as *mut u8, 32);
+            let shortest_len = inner_byte_slice.len().min(bytes.len());
+            inner_byte_slice[..shortest_len].copy_from_slice(&bytes[..shortest_len]);
+            inner_byte_slice.reverse();
+        }
+        #[cfg(target_endian = "big")]
+        for word in self.0.iter_mut() {
+            *word = word.swap_bytes();
+        }
+    }
+
+    /// Replaces the current value by interpreting the bytes in little endian order
+    pub fn copy_from_le_byte_slice(&mut self, bytes: &[u8]) {
+        // Same principle as in copy_from_be_byte_slice applies here
+
+        let inner_slice = self.0.as_mut_slice();
+        inner_slice.fill(0);
+        unsafe {
+            // SAFETY SLICES are contiguous
+            // (4 * 64) bits / 8 bits = 32
+            let inner_byte_slice =
+                std::slice::from_raw_parts_mut(inner_slice.as_mut_ptr() as *mut u8, 32);
+            let shortest_len = inner_byte_slice.len().min(bytes.len());
+            inner_byte_slice[..shortest_len].copy_from_slice(&bytes[..shortest_len]);
+        }
+        #[cfg(target_endian = "big")]
+        for word in self.0.iter_mut() {
+            *word = word.swap_bytes();
+        }
+    }
+
+    pub fn copy_to_le_byte_slice(&self, bytes: &mut [u8]) {
+        let inner_slice = self.0.as_slice();
+        unsafe {
+            // SAFETY SLICES are contiguous
+            // (4 * 64) bits / 8 bits = 32
+            let inner_byte_slice =
+                std::slice::from_raw_parts(inner_slice.as_ptr() as *const u8, 32);
+            let shortest_len = inner_byte_slice.len().min(bytes.len());
+            bytes[..shortest_len].copy_from_slice(&inner_byte_slice[..shortest_len]);
+        }
+        #[cfg(target_endian = "big")]
+        for sub_slice in bytes.chunks_mut(8) {
+            sub_slice.reverse();
+        }
+    }
+
+    pub fn copy_to_be_byte_slice(&self, bytes: &mut [u8]) {
+        let inner_slice = self.0.as_slice();
+        unsafe {
+            // SAFETY SLICES are contiguous
+            // (4 * 64) bits / 8 bits = 32
+            let inner_byte_slice =
+                std::slice::from_raw_parts(inner_slice.as_ptr() as *const u8, 32);
+            let shortest_len = inner_byte_slice.len().min(bytes.len());
+            bytes[..shortest_len].copy_from_slice(&inner_byte_slice[..shortest_len]);
+        }
+        #[cfg(target_endian = "big")]
+        for sub_slice in bytes.chunks_mut(8) {
+            sub_slice.reverse();
+        }
```

Comment:
same applies here, I don't know if I would not ask for 4 u64 on the C API side every time and be done with it for the moment, while I understand the flexibility of the bytes, it feels a bit silly (and if they want to they can do their own wrapper on the u64_words_to_u256 to manage smaller types)

I would probably see something like little_endian_u64_words_to_u256(word_0: u64, word_1: u64, word_2: u64, word_u64: 3, ...) and they can easily pass 0 where they want to pass zero ? Or was the byte slice thing a requirement from the block chain ? Also in the case of the byte slice I would probably require a 32 bytes slice or fixed size array of 32 bytes to avoid headaches here

Reply:
We could add a function that takes 4 u64, but I think as a common api to initialize from/to bytes
for example the go uint256 lib that I used, exposed a ToBigEndianBytes, which means that if we only have a function that takes 4u64, we would still need to reconstruct these u64 from the givens bytes, and this is what these function does

Reply:
up to you, I think I would require a fixed size byte slice to make things easier